floris/parallel_floris_model_2.py

```diff
@@ -0,0 +1,309 @@
+from __future__ import annotations
+
+import copy
+import warnings
+from pathlib import Path
+from time import perf_counter as timerpc
+
+import numpy as np
+import pandas as pd
+
+from floris.core import State
+from floris.floris_model import FlorisModel
+from floris.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR
+
+
+class ParallelFlorisModel(FlorisModel):
+    """
+    This class mimics the FlorisModel, but enables parallelization of the main
+    computational effort.
+    """
+
+    def __init__(
+        self,
+        configuration: dict | str | Path | FlorisModel,
+        interface: str | None = "multiprocessing",
+        max_workers: int = -1,
+        n_wind_condition_splits: int = 1,
+        return_turbine_powers_only: bool = False,
+        print_timings: bool = False
+    ):
+        """
+        Initialize the ParallelFlorisModel object.
+
+        Args:
+            configuration: The Floris configuration dictionary or YAML file.
+            The configuration should have the following inputs specified.
+                - **flow_field**: See `floris.simulation.flow_field.FlowField` for more details.
+                - **farm**: See `floris.simulation.farm.Farm` for more details.
+                - **turbine**: See `floris.simulation.turbine.Turbine` for more details.
+                - **wake**: See `floris.simulation.wake.WakeManager` for more details.
+                - **logging**: See `floris.simulation.core.Core` for more details.
+            interface: The parallelization interface to use. Options are "multiprocessing",
+               with possible future support for "mpi4py" and "concurrent"
+            max_workers: The maximum number of workers to use. Defaults to -1, which then
+               takes the number of CPUs available.
+            n_wind_condition_splits: The number of wind conditions to split the simulation over.
+               Defaults to the same as max_workers.
+            return_turbine_powers_only: Whether to return only the turbine powers.
+            print_timings (bool): Print the computation time to the console. Defaults to False.
+        """
+        # Instantiate the underlying FlorisModel
+        if isinstance(configuration, FlorisModel):
+            self.logger.warning(
+                "Received an instantiated FlorisModel, when expected a dictionary or path"
+                " to a FLORIS input file. Converting to dictionary to instantiate "
+                " the ParallelFlorisModel."
+            )
+            configuration = configuration.core.as_dict()
+        super().__init__(configuration)
+
+        # Save parallelization parameters
+        if interface == "multiprocessing":
+            import multiprocessing as mp
+            self._PoolExecutor = mp.Pool
+            if max_workers == -1:
+                max_workers = mp.cpu_count()
+            # TODO: test spinning up the worker pool at this point
```

Comment:
Still open?

Reply:
Still something I'd like to look into a bit further